pkg/storage/stores/shipper/indexshipper/gatewayclient/gateway_client.go

```diff
@@ -237,78 +237,109 @@ func (s *GatewayClient) QueryIndex(_ context.Context, _ *logproto.QueryIndexRequ
+func (s *GatewayClient) LabelNamesForMetricName(ctx context.Context, in *logproto.LabelNamesForMetricNameRequest) (*logproto.LabelResponse, error) {
 	var (
 		resp *logproto.LabelResponse
 		err  error
 	)
 	err = s.poolDo(ctx, func(client logproto.IndexGatewayClient) error {
-		resp, err = client.LabelNamesForMetricName(ctx, in, opts...)
+		resp, err = client.LabelNamesForMetricName(ctx, in)
 		return err
 	})
 	return resp, err
 }
 
-func (s *GatewayClient) LabelValuesForMetricName(ctx context.Context, in *logproto.LabelValuesForMetricNameRequest, opts ...grpc.CallOption) (*logproto.LabelResponse, error) {
+func (s *GatewayClient) LabelValuesForMetricName(ctx context.Context, in *logproto.LabelValuesForMetricNameRequest) (*logproto.LabelResponse, error) {
 	var (
 		resp *logproto.LabelResponse
 		err  error
 	)
 	err = s.poolDo(ctx, func(client logproto.IndexGatewayClient) error {
-		resp, err = client.LabelValuesForMetricName(ctx, in, opts...)
+		resp, err = client.LabelValuesForMetricName(ctx, in)
 		return err
 	})
 	return resp, err
 }
 
-func (s *GatewayClient) GetStats(ctx context.Context, in *logproto.IndexStatsRequest, opts ...grpc.CallOption) (*logproto.IndexStatsResponse, error) {
+func (s *GatewayClient) GetStats(ctx context.Context, in *logproto.IndexStatsRequest) (*logproto.IndexStatsResponse, error) {
 	var (
 		resp *logproto.IndexStatsResponse
 		err  error
 	)
 	err = s.poolDo(ctx, func(client logproto.IndexGatewayClient) error {
-		resp, err = client.GetStats(ctx, in, opts...)
+		resp, err = client.GetStats(ctx, in)
 		return err
 	})
 	return resp, err
 }
 
-func (s *GatewayClient) GetVolume(ctx context.Context, in *logproto.VolumeRequest, opts ...grpc.CallOption) (*logproto.VolumeResponse, error) {
+func (s *GatewayClient) GetVolume(ctx context.Context, in *logproto.VolumeRequest) (*logproto.VolumeResponse, error) {
 	var (
 		resp *logproto.VolumeResponse
 		err  error
 	)
 	err = s.poolDo(ctx, func(client logproto.IndexGatewayClient) error {
-		resp, err = client.GetVolume(ctx, in, opts...)
+		resp, err = client.GetVolume(ctx, in)
 		return err
 	})
 	return resp, err
 }
 
+func (s *GatewayClient) GetShards(
+	ctx context.Context,
+	in *logproto.ShardsRequest,
+) (*logproto.ShardsResponse, error) {
+	res := &logproto.ShardsResponse{}
+
+	if err := s.poolDo(ctx, func(client logproto.IndexGatewayClient) error {
+		streamer, err := client.GetShards(ctx, in)
+		if err != nil {
+			return errors.Wrap(err, "get shards")
+		}
+
+		// TODO(owen-d): stream currently unused because query planning doesn't expect a streamed response,
+		// but can be improved easily in the future by using a stream here.
+		for {
+			resp, err := streamer.Recv()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				return errors.WithStack(err)
+			}
+			res.Shards = append(res.Shards, resp.Shards...)
```

Comment:
The `append` is error prone here. `poolDo()` retries the job with the next instance in the replication set of a tenant in case of an error.

Reply:
nice catch, tyvm.